Complex gradcheck logic

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44744 Add tests for R to C cases
* #44444 [WIP] Update backward definition for more operators and reenable tests in test_ops.py
* **#43208 Complex gradcheck logic**

This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `sin`, `cos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:

- [ ]  Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)` (#44744 )

- [x]  Add back commented test in `common_methods_invocation.py`. ( #44444  )

- [ ]  Add more special case checking for complex gradcheck to make debugging easier.

- [x]  Update complex autograd note.

- [x]  disable complex autograd for operators not tested for complex. (#45461 )

- [x]  Re-enable tests in `test_ops.py` for complex dtype. ( #44444  )

- [ ]  Re-enable `TestGradCheckOverride.test_gradcheck` cc. @hameerabbasi (#45732 )

Differential Revision: [D23655088](https://our.internmc.facebook.com/intern/diff/D23655088)

anjali411 committed:
Complex autograd logic

[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex gradcheck logic"




[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -15,17 +15,26 @@ def zero_gradients(x):
             return None
-        return torch.zeros(input.nelement(), num_out, dtype=input.dtype)
+        shape = (input.nelement(), num_out)
+        # two dimensions to store dout_dx, dout_dy for an input x + yj
```
The comment above mentions that these contain wirtinger and conjugate wirtinger. These are not the same as derivatives wrt to real and imaginary parts respectively right?

anjali411 replied:
no forgot to update the comment after making some changes! updated

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -15,17 +15,26 @@ def zero_gradients(x):
+# is set to True. The extra dimension helps store both wirtinger and conjugate wirtinger derivative
+# required for gradcheck for complex functions.
+# For R -> C, wirtinger and conjugate wirtinger derivative are equal.
+def make_jacobian(input, num_out, extra_dim_if_complex=False):
```
What is the reason for this flag? Are there cases where you don't want to do this?

anjali411 replied:
yeah we don't wanna do this when this function is called in `get_analytical_jacobian`

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -53,38 +62,47 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
-        return r.detach().reshape(-1)
+    def update_jacobians(x, idx, d, d_idx, is_mkldnn=False):
+
+        def compute_gradient(delta=eps):
```
Do you need a default value here? You always provide it below.

anjali411 replied:
yup it's unnecessary. will remove it

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -53,38 +62,47 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
+        if x.is_complex():
+            ds_dy = compute_gradient(delta=(eps * 1j))
+            # conjugate wirtinger derivative
+            d[d_idx][:, 0] = 0.5 * (ds_dx + ds_dy * 1j)
```
nit: can you do this with a single indexing op?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -53,38 +62,47 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
 
     # TODO: compare structure
     for x_tensor, d_tensor in zip(x_tensors, j_tensors):
         is_complex = x_tensor.dtype.is_complex
```
This can be removed as well?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -286,7 +304,7 @@ def fail_test(msg):
             def fn(input):
                 return _as_tuple(func(*input))[i]
-            numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
+            numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)[0]
```
Old code I guess?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -302,16 +320,45 @@ def fn(input):
+
+        if out_is_complex:
+            # analytical vjp with v = 1.0j
+            analytical_with_imag_v, reentrant_with_imag_v, correct_grad_sizes_ \
```
`correct_grad_sizes_` is not a great name I think. why not `*_with_imag_v` as well?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -302,16 +320,45 @@ def fn(input):
+            analytical_with_imag_v, reentrant_with_imag_v, correct_grad_sizes_ \
+                = get_analytical_jacobian(tupled_inputs, o, nondet_tol=nondet_tol, v=1j)
+
+        if not correct_grad_sizes or (out_is_complex and not correct_grad_sizes_):
```
It would be nice to update the error message to tell which of the two does not have the correct size?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -302,16 +320,45 @@ def fn(input):
+        def checkIfNumericalAnalyticAreClose(a, n, error_str=''):
+            if not torch.allclose(a, n, rtol, atol):
+                return fail_test(error_str + 'Jacobian mismatch for output %d with respect to input %d,\n'
+                                 'numerical:%s\nanalytical:%s\n' % (i, j, n, a))
```
The value of `j` will be wrong here no?

anjali411 replied:
updated

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -302,16 +320,45 @@ def fn(input):
+                        a_with_imag_v = analytical_with_imag_v[j]
+                        checkIfNumericalAnalyticAreClose(a_with_imag_v, -1j * 0.5 * n + 1j * 0.5 * n.conj())
+
+        if not reentrant or (out_is_complex and not reentrant_with_imag_v):
```
Same here, would be nice to know which one is failing.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -302,16 +320,45 @@ def fn(input):
+                    # C -> R, C -> C
+                    checkIfNumericalAnalyticAreClose(a, n[:, :, 0] + n[:, :, 1].conj(),
+                                                     "Real value(s) of Conjugate Wirtinger derivative \
+                                                     (Re(ds/dx + ds/dy * 1j)) failed to compare equal.")
```
nit: missing a space at the end to concatenate properly with the other string no?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -302,16 +320,45 @@ def fn(input):
+                        # C -> C
+                        a_with_imag_v = analytical_with_imag_v[j]
+                        checkIfNumericalAnalyticAreClose(a_with_imag_v, -1j * n[:, :, 0] + 1j * n[:, :, 1].conj(),
+                                                         "Imaginary value(s) of Conjugate Wirtinger derivative \
```
You mention the conjugate wirtinger here and the wirtinger in the error message above but the formula is the same in both cases?

anjali411 replied:
I don't follow. Both error messages state that real/imag value of conjugate wirtinger derivative failed to compare equal.

albanD replied:
Misread the first one sorry.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -302,16 +320,45 @@ def fn(input):
+                    if o.is_complex():
+                        # R -> C
+                        a_with_imag_v = analytical_with_imag_v[j]
+                        checkIfNumericalAnalyticAreClose(a_with_imag_v, -1j * 0.5 * n + 1j * 0.5 * n.conj())
```
Missing a extra message here?

anjali411 replied:
added!

albanD reviewed (commented):
It would be nice to update the docstring of the `gradcheck` and `gradgradcheck` functions to explain what we test in the complex case.

anjali411 commented on test/test_autograd.py:
```diff
@@ -4644,18 +4645,18 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
 # and only run for floating point
 
-separate_complex_tests = ['log', 'log10', 'log1p', 'log2', 'reciprocal', 'tan']
+separate_complex_tests = ['view_as_real', 'real', 'imag']  # ['log', 'log10', 'log1p', 'log2', 'reciprocal', 'tan']
```
these tests should be added back after updating the formula. Added a todo

albanD replied:
Maybe tag them with your name or the PR number so that you can easily find them again (there are a LOT of "TODO" in the repo).

anjali411 commented on test/test_autograd.py:
```diff
@@ -4644,18 +4645,18 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
-                '__rmul__', '__rdiv__', 'sum', 'transpose', 'round', 'add', 'roll',
-                '__radd__', 'repeat', 'expand', 'mul', 'tanh', 'flip', 'fliplr', 'flipud',
-                'rot90'] + separate_complex_tests
+complex_list = ['t', 'view', 'reshape', 'reshape_as', 'view_as', 'roll', 'clone',
```
this is the list of functions whose vjp formula doesn't need to be updated. the commented tests should be added back after their formulas are updated based on the tf definition (adding a conj)

anjali411 commented on test/test_jit.py:
```diff
@@ -15398,7 +15398,7 @@ def add_autograd_test(
     # Disable complex tests
     # TODO: Add complex support for jit
-    if 'complex' in variant_name:
+    if 'complex' in variant_name or name in ['view_as_complex']:
```
this test runs with jit because it takes as input a real tensor. The test fails because it returns a complex tensor and complex tensors are not supported on jit

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1429,6 +1434,10 @@ def unpack_variables(args):
     'test_cdist',
 }
 
+EXCLUDE_METHOD_FOR_ATTRIBUTES = {
```
probably no unless we change a lot of these lists. this condition checks if name is attribute https://github.com/pytorch/pytorch/blob/master/test/test_autograd.py#L4756-L4757 and then uses the method call in `getattr` which does not work for `real` and `imag` unlike other attributes. This might also be addressed in new autograd test suite cc. @mruberry

albanD replied:
Is the issue that they are attributes and not functions?
In that case, maybe you just want to update `add_test` to do `hasattr(self_variable, name) and callable(getattr(self_variable, name))` ?

Or better still, just access the attribute instead of calling the function there so that you can properly test the attribute access

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 commented:
@albanD updated the PR based on your comments. Also:
1. fixed `test_autograd_simple_views_python` test failure.
2. update gradcheck documentation
3. Added TODOs tagged with my github handle
4. added more error messages.

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updated get_numerical_jacobian to return a jacobian with tensors containing an extra dimension whenever the input is complex. This is meant to store both wirtinger and conjugate wirtinger derivatives used later while comparing with analytical gradient.
2. Updated gradcheck logic to calculate analytical jacobian with grad_ouput vectors containing [...,1,...] and [...,1j,...]. Also added extra error message to help in debugging.
3. Updated vjp definition for `torch.view_as_real` and `torch.view_as_complex` which is used in `tensor.real` and `tensor.imag`. Also added tests for all these functions to verify correctness.
4. Disabled some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)
5. Added a test for `torch.conj`

[ghstack-poisoned]

anjali411 commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -605,10 +605,7 @@ void validate_outputs(
-    bool input_is_complex = isComplexType(c10::typeMetaToScalarType(metadata.options().dtype()));
-    bool grad_is_complex = isComplexType(grad.scalar_type());
-
-    TORCH_CHECK(isFloatingType(grad.scalar_type()) || (input_is_complex == grad_is_complex));
```
removed the check `(input_is_complex == grad_is_complex)` because this is not satisfied for `R -> C` functions

albanD replied:
It should be no? A real tensor should only have a real gradient, not a complex one right?

anjali411 replied:
yeah I confused it with `grad_out` so added it back again

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R-> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.
4. Disables some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R-> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.
4. Disables some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R-> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.
4. Disables some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R-> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.
4. Disables some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -4837,30 +4841,6 @@ def fn(*inputs):
-        self.assertEqual(x.grad, torch.full_like(x, 1 - 1j))
-
-    # remove this test after gradcheck support is added for non-holomorphic functions
-    def test_view_as_complex(self):
```
This was removed here but not added to the other tests? Should it stay here for now?

anjali411 replied:
added here: https://github.com/pytorch/pytorch/pull/43208/files#diff-9a6b441df38c89d996c08b92f1d086e9R522
it's not added in the complex list in `test_autograd.py` since this test needs as input a real valued tensor,

albanD commented on test/test_autograd.py:
```diff
@@ -4653,18 +4654,21 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
 # and only run for floating point
 
-separate_complex_tests = ['log', 'log10', 'log1p', 'log2', 'reciprocal', 'tan']
+# TODO: add the commented tests back after updating the formula based on tensorflow definition - @anjali411
```
nit: A convenient way to tag TODOs is to to `TODO(anjali411)` so that you search for that to get all the todos in the repo

albanD commented on test/test_ops.py:
```diff
@@ -89,33 +89,33 @@ def _gradgrad_test_helper(self, device, dtype, op, variant):
 
     # Tests that gradients are computed correctly
-    @dtypes(torch.double, torch.cdouble)
+    @dtypes(torch.double)
```
Why are these removed? Are you planning to add them bad later?
If so, there is a missing TODO, otherwise, you can just remove this and use `torch.double` as the dtype no?

cc @mruberry that might have some comments on the testing.

anjali411 replied:
added them back

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -211,6 +211,18 @@ Tensor mvlgamma_backward(Tensor grad, const Tensor & self, int64_t p) {
+  // only real valued gradient should be propagated for real tensors
+  if (!self.is_complex() && result.is_complex()) {
+    // R -> C
+    result = at::real(result);
```
Why is that the right thing to do?
Why not the sum of the real and imaginary part?

albanD replied:
Answer: The trick is that we implement the backward formula as if y was provided and was 0. And then take the real part of it. This gives the same result as if we derived the formula directly but it is simpler to write the overall code this way.

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -736,11 +736,11 @@
 - name: mul.Tensor(Tensor self, Tensor other) -> Tensor
-  self: grad * other
-  other: grad * self
+  self: grad * other.conj()
```
This will fail like the scalar version if self and other are not both real or both complex right?

anjali411 replied:
yeah updated to check for R-> C. this is being used in `sin` and `cos` gradgradcheck. will add more specific tests in follow up PR.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -54,37 +53,48 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
+            dL_dz_conj = 0.5 * (v.conjugate() * ds_dx + v * ds_dx.conj())
+            d[d_idx] = torch.real(dL_dz_conj)
+        else:
+            d[d_idx] = ds_dx
```
This should be multiplied by v no?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -54,37 +53,48 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
+            d[d_idx] = v.conjugate() * conj_w_d + v * w_d.conj()
+        elif ds_dx.is_complex():  # R -> C
+            dL_dz_conj = 0.5 * (v.conjugate() * ds_dx + v * ds_dx.conj())
+            d[d_idx] = torch.real(dL_dz_conj)
```
Why do you only take the real part here?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -42,7 +41,7 @@ def iter_tensors(x, only_requiring_grad=False):
                 yield result
 
-def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
+def get_numerical_jacobian(fn, input, target=None, eps=1e-3, v=1.0):
```
Missing doc on the new argument.

anjali411 replied:
updated

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -214,6 +224,9 @@ def gradcheck(
 
     The check between numerical and analytical gradients uses :func:`~torch.allclose`.
 
+    For functions with complex output, the gradients are computed using :attr:`grad_output`
```
I'm not sure what that means.
The matrix that you compare is the matrix containing all the conjugate wirtinger derivatives computed under the assumption that the overall function is C -> R no?
Also to keep this short, you can simply point to the autograd Note on complex if people want to know the definition of these terms.

anjali411 replied:
updated the description

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -308,23 +321,65 @@ def fn(input):
-                             'input and grad_output multiple times gives different values, '
-                             'although analytical gradient matches numerical gradient. '
-                             'The tolerance for nondeterminism was {}.'.format(nondet_tol))
+            not_reentrant_error()
```
This and below are missing some return statements no?

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -510,6 +515,12 @@ def method_tests():
         ('sinh', (), NO_ARGS, 'scalar', (True,)),
         ('cosh', (S, S, S), NO_ARGS, '', (True,)),
         ('cosh', (), NO_ARGS, 'scalar', (True,)),
+        ('conj', (S, S, S), NO_ARGS),
```
Do you want to test conj with complex inputs as well?

anjali411 replied:
it's being tested with both real and complex inputs since it's added in the list `complex_list`

albanD commented on test/test_autograd.py:
```diff
@@ -4837,30 +4841,6 @@ def fn(*inputs):
-    # remove this test after gradcheck support is added for non-holomorphic functions
-    def test_real(self):
-        x = torch.randn(3, 4, 5, dtype=torch.cdouble, requires_grad=True)
-        x.real.sum().backward()
```
Looking at `EXCLUDE_METHOD_FOR_ATTRIBUTES`, you actually skip this test where you access the attribute in the new tests you add. So this won't be tested in the new tests right?

anjali411 replied:
as per offline discussion, removed `EXCLUDE_METHOD_FOR_ATTRIBUTES` and `real` and `imag` should now be tested properly.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -308,23 +321,65 @@ def fn(input):
+                    a_with_imag_v = analytical_with_imag_v[j]
+                    n_with_imag_v = numerical_with_imag_v[j]
+                    checkIfNumericalAnalyticAreClose(a_with_imag_v, n_with_imag_v, j,
+                                                     "Imaginary value(s) of Conjugate Wirtinger derivative \
```
If this is just the imaginary part of the Conjugate Wirtinger, could you actually get, for the numerical jacobian, both this value and the one required line 363 below by taking the real and imaginary part of what is computed line 83: `conj_w_d = 0.5 * (ds_dx + ds_dy * 1j)` ?
This would remove one of the two calls to `get_numerical_jacobian`.

anjali411 replied:
The message mentioned earlier was incorrect. For `grad_out v = v1 + v2 j`,
``` 
ds_dJ_conj = 0.5 * [ v.conj() * (ds_dx + ds_dy * 1j) + v * (ds_dx - ds_dy * 1j) ]
                   = v1 * ds_dx.real + v2 * (ds_dx.imag + ds_dy.real) + v1 * (ds_dy.real + ds_dy.imag) * (1 + 1j)

For v = 1, we have v1 = 1, v2 = 0
ds_dJ_conj = ds_dx.real +(ds_dy.real + ds_dy.imag) * (1 + 1j)

For v = 1j, we have v1 = 0, v2 = 1
ds_dJ_conj = ds_dx.imag + ds_dy.real
```

The special cases clearly noticeable are:
```
- For  v=1, and f:`R->C`, ds_dJ_conj = ds_dx.real
- For v=1j, and f:`R->C`, ds_dJ_conj = 0
```

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R-> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.
4. Disables some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R-> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.
4. Disables some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)

[ghstack-poisoned]

anjali411 commented on tools/autograd/derivatives.yaml:
```diff
@@ -162,7 +162,7 @@
 
 - name: acos(Tensor self) -> Tensor
-  self: grad * -((-self * self + 1).rsqrt())
+  self: grad * -((-self * self + 1).rsqrt()).conj()
```
tested in `test_ops.py`

anjali411 commented on tools/autograd/derivatives.yaml:
```diff
@@ -251,7 +251,7 @@
 
 - name: asin(Tensor self) -> Tensor
-  self: grad * (-self * self + 1).rsqrt()
+  self: grad * (-self * self + 1).rsqrt().conj()
```
tested in `test_ops.py`

anjali411 commented on tools/autograd/derivatives.yaml:
```diff
@@ -900,7 +900,7 @@
 
 - name: rsqrt(Tensor self) -> Tensor
-  self: -0.5 * grad * result.pow(3)
+  self: -0.5 * grad * result.pow(3).conj()
```
currently gradgradcheck for `rsqrt` fails since it uses `pow_backward`. Will update `pow_backward` in follow up PRs and also add more extensive testing for it.

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R-> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.
4. Disables some tests for complex where an extra conjugation is needed (`sin`, `cos` etc.)

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `rsqrt`, `sin`, `cos`, `asin`, `acos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add gradgradcheck for `rsqrt` which would need updating the backward formula for `pow`.
2. Add more thorough tests for R -> C cases.
3. Add back commented test in `common_methods_invocation.py`.

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `rsqrt`, `sin`, `cos`, `asin`, `acos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add gradgradcheck for `rsqrt` which would need updating the backward formula for `pow`.
2. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
3. Add back commented test in `common_methods_invocation.py`.
4. Try and find more special case checking for complex gradcheck to make debugging easier.

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

This PR is bc breaking because after this PR, we will save intermediate variables for the backward calculation of mul.Scalar and mul.Tensor and in place modification of input tensor would lead to RuntimeError. 
```
        mu = torch.ones(1, requires_grad=True)
        x = torch.empty(1)
        loss = 0
        for i in range(3):
            x.detach_()
            x.copy_(mu + i)
            ft = torch.tensor([float(i)])
            multiplied = x * ft
            s = multiplied.sum()
            loss += s
        loss.backward()
```

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `rsqrt`, `sin`, `cos`, `asin`, `acos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add gradgradcheck for `rsqrt` which would need updating the backward formula for `pow`.
2. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
3. Add back commented test in `common_methods_invocation.py`.
4. Try and find more special case checking for complex gradcheck to make debugging easier.


[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

This PR is bc breaking because after this PR, we will save intermediate variables for the backward calculation of mul.Scalar and mul.Tensor and in place modification of input tensor would lead to RuntimeError. 
```
        mu = torch.ones(1, requires_grad=True)
        x = torch.empty(1)
        loss = 0
        for i in range(3):
            x.detach_()
            x.copy_(mu + i)
            ft = torch.tensor([float(i)])
            multiplied = x * ft
            s = multiplied.sum()
            loss += s
        loss.backward()
```

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `rsqrt`, `sin`, `cos`, `asin`, `acos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add gradgradcheck for `rsqrt` which would need updating the backward formula for `pow`.
2. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
3. Add back commented test in `common_methods_invocation.py`.
4. Try and find more special case checking for complex gradcheck to make debugging easier.


[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `rsqrt`, `sin`, `cos`, `asin`, `acos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add gradgradcheck for `rsqrt` which would need updating the backward formula for `pow`.
2. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
3. Add back commented test in `common_methods_invocation.py`.
4. Try and find more special case checking for complex gradcheck to make debugging easier.


[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `rsqrt`, `sin`, `cos`, `asin`, `acos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add gradgradcheck for `rsqrt` which would need updating the backward formula for `pow`.
2. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
3. Add back commented test in `common_methods_invocation.py`.
4. Try and find more special case checking for complex gradcheck to make debugging easier.


[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `rsqrt`, `sin`, `cos`, `asin`, `acos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add gradgradcheck for `rsqrt` which would need updating the backward formula for `pow`.
2. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
3. Add back commented test in `common_methods_invocation.py`.
4. Try and find more special case checking for complex gradcheck to make debugging easier.


[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `sin`, `cos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
2. Add back commented test in `common_methods_invocation.py`.
3. Add more special case checking for complex gradcheck to make debugging easier.


[ghstack-poisoned]

ezyang commented on test/test_autograd.py:
```diff
@@ -4759,7 +4763,13 @@ def fn(*inputs):
                         if hasattr(self_variable, name):
-                            output_variable = getattr(self_variable, name)(*args_variable, **kwargs_variable)
+                            attribute_result = getattr(self_variable, name)
+                            if callable(attribute_result):
```
lol

albanD replied:
Is that a good lol or a bad lol? 😄

ezyang commented on torch/autograd/gradcheck.py:
```diff
@@ -62,30 +63,48 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
+            r = (outb - outa) / (2 * eps)
+            return r.detach().reshape(-1)
+
+        ds_dx = compute_gradient(eps)
```
Helpful to cite the boedekker here!

ezyang replied:
I'm guessing that above `s` is referred to as `outa` and `outb`

anjali411 replied:
makes sense! will do

ezyang commented on torch/autograd/gradcheck.py:
```diff
@@ -62,30 +63,48 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
+            return r.detach().reshape(-1)
+
+        ds_dx = compute_gradient(eps)
+        if x.is_complex():
```
sooo... is this C -> C?

anjali411 replied:
> ```diff
> +    def __mul__(self, other):
> +        return self.__torch_function__(torch.mul, (Wrapper,), (self, other))
> ```

Then it fails with:

```
======================================================================
ERROR: test_gradcheck (__main__.TestGradCheckOverride)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_overrides.py", line 775, in test_gradcheck
    gradcheck(torch.add, (a, b), raise_exception=False)
  File "/home/chourdiaanjali/pytorch2/torch/autograd/gradcheck.py", line 461, in gradcheck
    outputs_to_check = [[torch._C._functions.UndefinedGrad()(o) for o in _differentiable_outputs(func(*tupled_inputs))]]
  File "/home/chourdiaanjali/pytorch2/torch/autograd/gradcheck.py", line 461, in <listcomp>
    outputs_to_check = [[torch._C._functions.UndefinedGrad()(o) for o in _differentiable_outputs(func(*tupled_inputs))]]
TypeError: argument 0 is not a Variable

----------------------------------------------------------------------
Ran 1 test in 0.097s
```

anjali411 replied:
C -> C, C -> R. also added a comment to make it more clear

hameerabbasi commented:
Can you try this patch (output of `git diff`):

```diff
diff --git a/test/test_overrides.py b/test/test_overrides.py
index e8d393d651..668ad2faa1 100644
--- a/test/test_overrides.py
+++ b/test/test_overrides.py
@@ -697,6 +697,9 @@ class Wrapper:
     def __sub__(self, other):
         return self.__torch_function__(torch.sub, (Wrapper,), (self, other))
 
+    def __mul__(self, other):
+        return self.__torch_function__(torch.mul, (Wrapper,), (self, other))
+
     def __truediv__(self, other):
         return self.__torch_function__(torch.true_divide, (Wrapper,), (self, other))
 
```

ezyang reviewed (approved):
Subtle stuff. Thanks for taking the time to work this all out.

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `sin`, `cos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
2. Add back commented test in `common_methods_invocation.py`.
3. Add more special case checking for complex gradcheck to make debugging easier.
4. Update complex autograd note.
5. disable complex autograd for operators not tested for complex.


[ghstack-poisoned]

anjali411 commented:
> Can you try this patch (output of `git diff`):
> 
> ```diff
> diff --git a/test/test_overrides.py b/test/test_overrides.py
> index e8d393d651..668ad2faa1 100644
> --- a/test/test_overrides.py
> +++ b/test/test_overrides.py
> @@ -697,6 +697,9 @@ class Wrapper:
>      def __sub__(self, other):
>          return self.__torch_function__(torch.sub, (Wrapper,), (self, other))
>  
> +    def __mul__(self, other):
> +        return self.__torch_function__(torch.mul, (Wrapper,), (self, other))
> +
>      def __truediv__(self, other):
>          return self.__torch_function__(torch.true_divide, (Wrapper,), (self, other))
>  
> ```

after this patch, the test fails with:
```
======================================================================
ERROR: test_gradcheck (__main__.TestGradCheckOverride)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_overrides.py", line 775, in test_gradcheck
    gradcheck(torch.add, (a, b), raise_exception=False)
  File "/home/chourdiaanjali/pytorch2/torch/autograd/gradcheck.py", line 461, in gradcheck
    outputs_to_check = [[torch._C._functions.UndefinedGrad()(o) for o in _differentiable_outputs(func(*tupled_inputs))]]
  File "/home/chourdiaanjali/pytorch2/torch/autograd/gradcheck.py", line 461, in <listcomp>
    outputs_to_check = [[torch._C._functions.UndefinedGrad()(o) for o in _differentiable_outputs(func(*tupled_inputs))]]
TypeError: argument 0 is not a Variable

----------------------------------------------------------------------
Ran 1 test in 0.097s
```

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `sin`, `cos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
2. Add back commented test in `common_methods_invocation.py`.
3. Add more special case checking for complex gradcheck to make debugging easier.
4. Update complex autograd note.
5. disable complex autograd for operators not tested for complex.


[ghstack-poisoned]

hameerabbasi commented:
Reversion is best for now I think.

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `sin`, `cos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
2. Add back commented test in `common_methods_invocation.py`.
3. Add more special case checking for complex gradcheck to make debugging easier.
4. Update complex autograd note.
5. disable complex autograd for operators not tested for complex.


[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -733,11 +733,11 @@
 
 - name: mul.Scalar(Tensor self, Scalar other) -> Tensor
-  self: grad * other
+  self: mul_scalar_backward(grad, other, self.scalar_type())
```
I think you can remove the `mul_scalar_backward` to reduce code duplication and just call `mul_tensor_backward(grad, at::scalar_to_tensor(other), self.scalar_type())`.

anjali411 replied:
good idea! thanks

anjali411 commented on test/test_overrides.py:
```diff
@@ -757,51 +760,51 @@ def test_wrapper(self):
-            torch.autograd.grad,
-            torch.add,
-        })
+# TODO(@anjali411): re-enable this test
```
disabling this for this PR. @hameerabbasi will re-enable this in a follow up PR

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -215,6 +238,13 @@ def gradcheck(
 
     The check between numerical and analytical gradients uses :func:`~torch.allclose`.
 
+    For complex functions, no notion of Jacobian exists. Gradcheck verifies if the numerical and
```
This looks great!

albanD reviewed (commented):
This looks quite good now!

Mostly removing extra code and variable naming as pointed out by Ed.

ngimel commented on torch/autograd/gradcheck.py:
```diff
@@ -62,30 +63,52 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
+            x[idx] = orig + delta
+            outb = fn_out()
+            x[idx] = orig
+            r = (outb - outa) / (2 * eps)
```
so this works as expected only if delta is `eps` or `j*eps`? Seems pretty fragile.

anjali411 replied:
for both cases. It's because for the case of complex functions with input x + yj (where both x and y are real) and output s, we are interested in calculating ds/dx and ds/dy. ds/dx and ds/dy are further used to calculate wirtinger and conjugare wirtinger derivatives.

ngimel replied:
I understand, but nothing prevents the caller from using some other argument, not `eps` or `j*eps`, in which case the result would be unexpected.

anjali411 replied:
yes you are right. added some comments and an assert based on your suggestion https://github.com/pytorch/pytorch/pull/43208/files#diff-0687821017cb7b20f80a3a8ab5daca97R7

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `sin`, `cos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
2. Add back commented test in `common_methods_invocation.py`.
3. Add more special case checking for complex gradcheck to make debugging easier.
4. Update complex autograd note.
5. disable complex autograd for operators not tested for complex.

Differential Revision: [D23655088](https://our.internmc.facebook.com/intern/diff/D23655088)

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `sin`, `cos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:
1. Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)`
2. Add back commented test in `common_methods_invocation.py`.
3. Add more special case checking for complex gradcheck to make debugging easier.
4. Update complex autograd note.
5. disable complex autograd for operators not tested for complex.
6. Re-enable tests in `test_ops.py` for complex dtype.
7. Re-enable `TestGradCheckOverride.test_gradcheck` cc. @hameerabbasi 

Differential Revision: [D23655088](https://our.internmc.facebook.com/intern/diff/D23655088)

[ghstack-poisoned]

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `sin`, `cos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:

- [ ]  Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)` (#44744 )

- [ ]  Add back commented test in `common_methods_invocation.py`. ( #43208 )

- [ ]  Add more special case checking for complex gradcheck to make debugging easier.

- [ ]  Update complex autograd note.

- [ ]  disable complex autograd for operators not tested for complex.

- [ ]  Re-enable tests in `test_ops.py` for complex dtype. ( #43208 )

- [ ]  Re-enable `TestGradCheckOverride.test_gradcheck` cc. @hameerabbasi 

Differential Revision: [D23655088](https://our.internmc.facebook.com/intern/diff/D23655088)

[ghstack-poisoned]

anjali411 commented on torch/autograd/gradcheck.py:
```diff
@@ -62,30 +63,57 @@ def get_numerical_jacobian(fn, input, target=None, eps=1e-3):
+        # or pure imaginary delta
+        def compute_gradient(delta):
+            # we currently assume that the norm of delta equals eps
+            assert(delta == eps or delta == (eps * 1j))
```
cc. @ngimel

anjali411 committed:
Update on "Complex gradcheck logic"


This PR adds gradcheck for complex. The logic used for complex gradcheck is described in Section 3.5.3 here: https://arxiv.org/pdf/1701.00392.pdf

More concretely, this PR introduces the following changes:
1. Updates get_numerical_jacobian to take as input a scalar value for vector (v). Adds gradcheck logic for C -> C, C-> R, R -> C. For R -> C functions, only the real value of gradient is propagated.
2. Adds backward definition for `torch.complex` and also adds a test to verify the definition added.
3. Updates backward for `mul`, `sin`, `cos`, `sinh`, `cosh`.
4. Adds tests for all `torch.real`, `torch.imag`, `torch.view_as_real`, `torch.view_as_complex`, `torch.conj`.

Follow up tasks:

- [ ]  Add more thorough tests for R -> C cases. Specifically, add R->C test variants for functions. for e.g., `torch.mul(complex_tensor, real_tensor)` (#44744 )

- [ ]  Add back commented test in `common_methods_invocation.py`. ( #43208 )

- [ ]  Add more special case checking for complex gradcheck to make debugging easier.

- [ ]  Update complex autograd note.

- [ ]  disable complex autograd for operators not tested for complex.

- [ ]  Re-enable tests in `test_ops.py` for complex dtype. ( #43208 )

- [ ]  Re-enable `TestGradCheckOverride.test_gradcheck` cc. @hameerabbasi 

Differential Revision: [D23655088](https://our.internmc.facebook.com/intern/diff/D23655088)

[ghstack-poisoned]

anjali411 commented:
@hameerabbasi this PR has been merged so would you like to create a follow-up PR to re-enable `TestGradCheckOverride.test_gradcheck`?

hameerabbasi commented:
I'll work on that soon, thanks.

anjali411 commented:
> I'll work on that soon, thanks.

Hi @hameerabbasi just checking in if you are still planning to work on re-enabling `TestGradCheckOverride.test_gradcheck`

anjali411 commented:
@hameerabbasi gentle ping

hameerabbasi commented:
There's a PR up that I submitted yesterday, #45732. Requested your review on it.

anjali411 commented:
> There's a PR up that I submitted yesterday, #45732. Requested your review on it.

thanks @hameerabbasi it looks great. I totally missed your PR before, but looks like @albanD is already reviewing it :D
